packages/core/types/index.js

```diff
@@ -299,6 +302,42 @@ export interface Asset extends BaseAsset {
   +stats: Stats;
 }
 
+// TODO: figure out naming
+export interface IConfig {
+  searchPath: FilePath;
+  env: Environment;
+  options: ParcelOptions;
+  resolvedPath: ?FilePath;
+  // $FlowFixMe could be anything
+  result: ?any;
+  resultHash: ?string;
+  includedFiles: Set<FilePath>;
+  watchGlob: ?Glob;
+  devDeps: Map<PackageName, ?string>;
+  pkg: ?PackageJSON;
+
+  setResolvedPath(filePath: FilePath): void;
+  // $FlowFixMe could be anything
+  setResult(result: any): void; // TODO: fix
+  setResultHash(resultHash: string): void;
```

Comment:
Did you consider an API where you the result directly from the plugin, instead of mutating the passed object? We could also automatically compute the hash from the result in core, unless there's a case where leaving that to the plugin is required.

```javascript
loadConfig({config}) {
  return {
    filePath: '...', // instead of setResolvedPath
    config: {...}    // instead of setResult
  };
}
```

I'm ok with either, just wanted to ensure we considered it.

Reply:
Yeah, that's something I've discussed with @wbinnssmith a bit. One reason I was leaning this way is that it is more consistent with how we pass `Asset` to the other methods. It also alleviates some work with methods like `getConfig` that both help with resolving/parsing config and also adding whichever file was resolve to included files.

> We could also automatically compute the hash from the result in core, unless there's a case where leaving that to the plugin is required.

Yeah, my thinking was it would be nice to be flexible because it's possible a resolved config might contain more information than is relevant for invalidation. One example where we're actually using it right now is the parcel config. For invalidation of a pipeline transformation, we only care about one pipeline, not the entire config. We could accomplish the same thing in other ways for Parcel, but it seemed like it might be useful for plugins. For instance, plugins might be using information from package.json, but they only care a single field. (That reminds me the babel transformer should probably invalidate if the `source` field changes)

Reply:
With transformers we kinda actually support both: there is that `TransformerResult` type which has fields like `code` and `map` that you can return instead of mutating the asset. We found this to be challenging to deal with when eg adding dependencies, and passing assets into util functions, and mutation was easier there. The difference here is that Assets start out fully hydrated (they have input code etc.), whereas config starts out as an empty object. It just felt a little strange. I'm not totally against what's here, just curious what it would look like another way.

I think your usecase with hashes is a good one to allow control to plugins, but I think we should make it optional. i.e. if no hash is set by the plugin, we could set one automatically based on the result. Perhaps the `shouldReload` setting could also set the hash to something random?

Reply:
I haven't had the time to explore this yet. Is this a blocker for the PR or could we follow up on it?